File: ios/Classes/TiGooglemapsRendererProxy.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2017 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsRendererProxy.h"
#import "TiGooglemapsView.h"
#import "TiGooglemapsViewProxy.h"

@implementation TiGooglemapsRendererProxy

- (GMUGeometryRenderer *)renderer
{
  if (_renderer == nil) {
    id mapView = [self valueForKey:@"mapView"];
    id file = [self valueForKey:@"file"];

    ENSURE_TYPE(mapView, TiGooglemapsViewProxy);
    ENSURE_TYPE(file, NSString);

    NSURL *url = [TiUtils toURL:file proxy:self];

    if (url == nil) {
      [self throwException:@"Invalid file provided" subreason:@"The file provided could not be found." location:CODELOCATION];
      return;
    }

    GMUKMLParser *parser = [[GMUKMLParser alloc] initWithURL:url];
    [parser parse];

    _renderer = [[GMUGeometryRenderer alloc] initWithMap:[(TiGooglemapsView *)[(TiGooglemapsViewProxy *)mapView view] mapView]
                                              geometries:[parser placemarks]
                                                  styles:[parser styles]];
  }

  return _renderer;
}

- (void)render:(id)unused
{
  [[self renderer] render];
}

- (void)clear:(id)unused
{
  [[self renderer] clear];
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsRendererProxy.swift
/**
 * Axway Titanium
 */

import Foundation
import TitaniumKit
import GoogleMaps
import GoogleMapsUtils

@objc(TiGooglemapsRendererProxy)
public class TiGooglemapsRendererProxy: TiProxy {
  private var geometryRenderer: GMUGeometryRenderer?
  
  private func runOnMainThread(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, false)
    }
  }
  
  private func rendererInstance() -> GMUGeometryRenderer? {
    if let renderer = geometryRenderer {
      return renderer
    }
    
    guard let mapViewProxy = value(forKey: "mapView") as? TiGooglemapsViewProxy,
          let tiView = mapViewProxy.mapView(),
          let file = value(forKey: "file") as? String,
          let url = TiUtils.toURL(file, proxy: self) else {
      return nil
    }
    
    let parser = GMUKMLParser(url: url)
    parser.parse()
    
    let renderer = GMUGeometryRenderer(map: tiView.mapView,
                                       geometries: parser.placemarks,
                                       styles: parser.styles)
    geometryRenderer = renderer
    return renderer
  }
  
  @objc public func setMapView(_ value: Any?) {
    replaceValue(value ?? NSNull(), forKey: "mapView", notification: false)
    geometryRenderer = nil
  }
  
  @objc public func setFile(_ value: Any?) {
    replaceValue(value ?? NSNull(), forKey: "file", notification: false)
    geometryRenderer = nil
  }
  
  @objc public func render(_ unused: Any?) {
    runOnMainThread {
      self.rendererInstance()?.render()
    }
  }
  
  @objc public func clear(_ unused: Any?) {
    runOnMainThread {
      self.rendererInstance()?.clear()
    }
  }
}
